### DISCUSSION TITLE
Long Press - potential feature request

### DISCUSSION THREAD

**Discussion by guttermonk** [bulletmark/libinput-gestures#367] [Ideas]
Would it be possible to add a Long Press gesture that would work similar to Hold On but would trigger on a longer press (instead of a tap)? Perhaps the time of the longer press could be set in the config. For example, you could put 3000 in the following command for a 3000ms press:

`gesture long press 1 3000 hyprctl dispatch workspace 1`

Just wanted to throw the idea out there. Seems like it would be an easy lift, since it would be a derivation of Hold On. Thanks for the great work on this program.

**Comment by bulletmark** [comment]
I can imagine this could be added but it would be awkward to implement and define in the configuration ( I would probably just extend `hold on` to `hold on-3` as optional hold + 3 secs rather than a whole new gesture type) and I also think it would be awkward to use because the user would have to guess the timeouts. Since you can optionally add a finger count (1 to 4) to "hold on" gestures then you have 4 actions available. It is easy to use a different number of fingers but coordinating that and subtle timing guesses would be too awkward and mistake-prone I think, unless you used very large timeouts which would just be too slow.

**Comment by guttermonk** [comment]
I would think most people know how long a long press is roughly, and I like the idea of combining it with the Hold On gesture. I wouldn't think the time you press would need to be exact. Really the time is just there to distinguish it from a regular tap, and could be treated as a greater than or equal to. So if you set `hold on-2` and pressed for two seconds or more, then the gesture would be picked up. 

Plus it would open up a whole new set of gestures, and allow me correlate the number of fingers to the workspace number, as so:
```
gesture hold on-2 1 hyprctl dispatch workspace 1
gesture hold on-2 2 hyprctl dispatch workspace 2
gesture hold on-2 3 hyprctl dispatch workspace 3
gesture hold on-2 4 hyprctl dispatch workspace 4
```
Currently, Hold On is only good for 3 or 4 fingers. Trying to use it with 1 or 2 fingers interferes with the left and right click. A long press feature could open up more possibilities. Anyway, just a thought.

**Comment by bulletmark** [comment]
I've implemented this in a new branch [hold-delay](https://github.com/bulletmark/libinput-gestures/tree/hold-delay) so can you please try that. It is implemented with fixed options for the times (1 to 10 secs) because implementation is easier that way. Read the new section on [Hold Gestures](https://github.com/bulletmark/libinput-gestures/tree/hold-delay#hold-gestures).

**Comment by guttermonk** [comment]
Sorry for the delay, it took me a while to get some help figuring out how to install a branch on NixOS. The hold-delay branch is working well for my use case. I set the delay to one second, but the delay still feels slightly longer than what I'm used to. I have to count one second in my head instead of going off of feel. I wonder if this could work with fractions of a second.

**Reply by bulletmark** [reply]
As I said it is much easier for me to implement this as fixed times, 1 to 10. I can't see how anybody could manually gauge times to a half second resolution with any reliability? So I don't think you want 0.5 sec resolution  but you are actually saying the **first** 1 sec is too slow. So just use zero delay. I.e. If you feel that `on-1` is too slow then just use `on-0`, i.e. `on` without a delay. E.g. for you use case then configure something like:

```
gesture hold on 3 hyprctl dispatch workspace 1
gesture hold on 4 hyprctl dispatch workspace 2
gesture hold on-1 3 hyprctl dispatch workspace 3
gesture hold on-1 4 hyprctl dispatch workspace 4
```

So:

- a very short 3 finger hold would activate `workspace 1` and a slightly longer 3 finger hold would activate `workspace 3`
- a very short 4 finger hold would activate `workspace 2` and a slightly longer 4 finger hold would activate `workspace 4`.

Or are you really asking for a `on-0.5` in between `on` and `on-1`?

**Reply by guttermonk** [reply]
Yes, I think being able to enter the delay in milliseconds or fractions of a second would be a minor improvement. A full second works, but it does feel a little slow compared to a long press on a smartphone (ios or android). Anyway, just trying to pass along some constructive feedback. Overall, I'm very excited to see this new feature added. As an added benefit, I'm able to bounce between two workspaces if I enter the same gesture twice in a row, which is extremely convenient. Kudos!

On a side-note, [the readme for Hold On](https://github.com/bulletmark/libinput-gestures?tab=readme-ov-file#hold-gestures) says that it is "actioned with 1 or more fingers," but when I tried adding `gesture hold on-1 5 hyprctl dispatch workspace 5` to my config it wouldn't respond. Can this only be actioned with 1-4 fingers?

**Reply by bulletmark** [reply]
I just pushed a commit on that hold-delay branch so that if you don't have a hold configured then you will see the `on-N` delay output in the debug log. E.g. comment out your hold commands and then run `libinput-gestures -d` and play with different hold timeouts using your fingers and you will see the `on-N` for N secs that would apply. Please do that and see if you can really distinguish between `on` and `on-1` so well that you think you could configure use `on-0.5` in between them. I try this and there is no way I could use that. A `on` hold is fast and a `on-1` is only barely slightly slower.

**Reply by guttermonk** [reply]
I just played with the -d and -r flags, and my taps registered around 1/10th of a second while what I thought would be a comfortable long-press registered between .5-.7 seconds. Human fast twitch muscles have a 40-90 millisecond contraction time on average, and musicians regularly deliver at those resolutions. If you think it's not possible to consistently tell the difference between .1, .5, or 1 second, then explain how [Danny Carey can do this](https://www.youtube.com/watch?v=FssULNGSZIA). I'm no Danny Carey, but just saying.

**Comment by bulletmark** [comment]
Regarding your question about 5 finger holds. I stated 1 to 5 fingers in my README because the libinput [hold gesture documentation](https://wayland.freedesktop.org/libinput/doc/latest/gestures.html#hold-gestures) does not explicitly preclude 5 finger holds and I think it depends on your touchpad. My touchpad does not report 5 finger holds either but there may be some out there that do so it is better that I don't limit it.

**Comment by bulletmark** [comment]
BTW, you say "A full second works, but it does feel a little slow compared to a long press on a smartphone (ios or android)." I just want to be clear that you can still use the original `on`. I.e. a hold without any extra delay. My question to you is if you use both `on` and `on-1` in your configuration could you really reliably discern another one between them?

**Reply by guttermonk** [reply]
I understand I could use the original `on` with no delay. I'm currently using both. Here are the `hold on` gestures I've added so far:
```
      gesture hold on 3 alacritty
      gesture hold on 4 hyprctl dispatch exec "kitty -e yazi"
      gesture hold on-1 1 hyprctl dispatch workspace 1
      gesture hold on-1 2 hyprctl dispatch workspace 2
      gesture hold on-1 3 hyprctl dispatch workspace 3
      gesture hold on-1 4 hyprctl dispatch workspace 4
```
If I could use milliseconds or fractions of a second with `hold on`, I would likely convert all of the on-1 entries to on-666 or on-0.66, depending on how it was set up. As to whether I could decern between on and on-666/on-0.66, I don't think that would be an issue since I'm able to discernibly use a tap and long press on my smartphone without any issue. That said, I don't have any data to how long a long-press is on android or ios. For all I know, it could be a full second too - I'll play with the -d flag and see if there are consistent delay times.

**Comment by bulletmark** [comment]
I have changed the hold-delay branch to work as you prefer. Now you can specify any arbitary time as an integer or float value in decimal seconds. The separator is changed to "+". E.g `on`, `on+0.55`, `on+3.4`, etc. You can specify any number and any times as you like, with or without a finger count.

Comment out your hold gestures and run `libinput-gestures -d` to see the times printed out for the durations you hold. That is an easy way to determine what you should configure.

**Reply by guttermonk** [reply]
This works great. I love it! Best gestures program out there. Thank you for adding this feature.

**Comment by bulletmark** [comment]
I have released new version 2.77 which includes this change.
